pkg/api/rest/delete.go

```diff
@@ -89,5 +92,13 @@ func BeforeDelete(strategy RESTDeleteStrategy, ctx api.Context, obj runtime.Obje
 	now := unversioned.NewTime(unversioned.Now().Add(time.Second * time.Duration(*options.GracePeriodSeconds)))
 	objectMeta.DeletionTimestamp = &now
 	objectMeta.DeletionGracePeriodSeconds = options.GracePeriodSeconds
+	// If it'a first graceful deletion we're going to update the object first,
+	// setting DeletionTimestamp to something != nil. Object that's being deleted shouldn't
+	// take any nontrivial actions, hence it's behavior changes. Thus we need to bump object's
+	// Generation (if set). This handles generation bump during graceful deletion. For bump
+  // caused by the finalizer-driven cascading deletion go to pkg/registry/generic/registry/store.go
+	if objectMeta.Generation > 0 {
```

Comment:
@smarterclayton do you foresee the damage caused by bumping generation twice? Controllers will just acknowledge it sees the latest one, so they don't care how many times the generation get bumped.